Add `detection_window` handling

TomWagg opened the pull request:
Hey @moeyensj! I've no idea if this will actually be useful to you but I was waiting for some code to run so I figured it wouldn't hurt to port across the stuff I made to a PR for difi.

The main change is to edit the `nightly_linkages` metric to additionally consider a `detection_window` of 15 nights and only mark something as findable if the other criteria are satisfied within this time frame as @mjuric suggested would be useful. This should close #33.

Additionally, I was finding it useful to know the night on which the object first became findable and to that end I added `difi.utils.calcFirstFindableNight`. I imagine this is probably not the right location but I figured you could play with it if you fancied. 

Note that all of the calculations necessary to get this first findable night are now already done in `_findNightlyLinkages` but it didn't seem like it was set up to return values in addition to obs_ids so I figured I could make it a postprocessing function instead. But perhaps it would be useful to include by default?

Hope this is helpful 😄

moeyensj commented:
Hey @TomWagg! 

Thanks for the PR, this is awesome and I'm more than happy to merge it once it's all working (looks like I have some bit rot that I need to fix with the Docker CI workflow). The calcFirstFindableNight function is a great addition (`utils` works as the appropriate location). 

One thing that might be worth considering is to generalize the `detection_window` parameter a level higher and have it be a parameter that gets applied in `analyzeObservations` and not just inside the tracklet metric. The most common use-case that I would encounter is to run both the `nightly_linkages` and `min_obs` metrics by writing an external for-loop that passes a sliding window of observations to the `analyzeObservations` function. The length of that sliding window would be equivalent to the `detection_window` parameter in this PR. I would then track each window by adding a window_id to the output data products. 

As an implementation, I think all you would need to do is write a for-loop inside the `analyzeObservations` function that splits the observations into windows, and then feeds those windows to the findability metrics. That way each metric benefits from the `detection_window` parameter and not just the tracklet one. I'd imagine this would simplify tests a little as well but I haven't given it that much thought. Does that make sense? What do you think?

TomWagg commented:
Ah yeah sure that makes sense. I hadn't realised that the detection window would apply in the `min_obs` case as well so I was just assuming you could use it as a `**metric_kwarg` for `nightly_linkages`.

This definitely seems like a good plan to me though. I also think I could make it a bit faster by removing objects once they get found in a particular window (and could then naturally set `first_findable_night` at that time). Although I guess that means that the `obs_ids` column would then only include observations from the window in which an object is detected - this seems like it would be okay though?

Alternatively I could keep every object and just parallelise the loop with a multiprocessing pool?

TomWagg commented:
Hey @moeyensj, I finally got round to doing this. I've removed the detection_window stuff from metric-specific code and it now works like this instead:

- `analyzeObservations` now has 2 new parameters
    - `detection_window` (how long is the window)
    - `ignore_after_detected` (whether to ignore objects in subsequent windows after it gets detected in one`
- If the user specifies a `detection_window` that's longer than the number of nights in the observations then I loop over the potential detection windows and get findable observations for each. I then concatenate these and return them as `findable_observations`
- I also added a column `window_first_night` in `findable_observations` to track the first night of the window in which it got marked as findable

Question for you: **Could we parallelise the loop over detection windows?** I'm not sure how we'd pass it to a pool since there are lots of changing parameters (I'm only used to it when one thing changes). (Also this obviously can only be done with `ignore_after_detected=False`).
